twoliter/embedded/rpm2img

```diff
@@ -240,6 +240,37 @@ INVENTORY_DATA="$(jq --arg PKG_PREFIX "bottlerocket-" \
 INVENTORY_DATA="$(jq --slurp 'sort_by(.Name)' <<< "${INVENTORY_DATA}" | jq '{"Content": .}')"
 printf "%s\n" "${INVENTORY_DATA}" > "${ROOT_MOUNT}/usr/share/bottlerocket/application-inventory.json"
 
+elf_note() {
+  local input output size offset uint32_t
+  input="${1:?}"
+  output="${2:?}"
+  name="Bottlerocket"
+  name_len="$(( ${#name} + 1 ))"
+  input_len="$(( ${#input} + 1 ))"
```

Comment:
This isn't quite right because the name has to be padded to a four-byte boundary ([source](https://www.netbsd.org/docs/kernel/elf-notes.html)) including the null terminator.

If I try this as-is then readelf complains:
```
Displaying notes found in: .note.bottlerocket.variant
  Owner                Data size 	Description
readelf: ./target/release/legolas: Warning: note with invalid namesz and/or descsz found at offset 0
readelf: ./target/release/legolas: Warning:  type: 0, namesize: 0xd, descsize: 0x14, alignment: 4
```

Reply:
A bit tricky because we need to write out runs of null bytes to achieve the padding, which is hard in bash.

I ended up with this:
```
elf_note() {
  local input output name
  local input_len input_pad input_fmt
  local name_len name_pad name_fmt
  local var val len pad fmt n sz
  input="${1:?}"
  output="${2:?}"
  name="Bottlerocket"

  # Handle null-termination, padding, and formatting.
  for var in name input ; do
    val="${!var}"
    len="${var}_len"
    pad="${var}_pad"
    fmt="${var}_fmt"
    # Add one byte for null terminator.
    printf -v "${len}" "$(( "${#val}" + 1 ))"
    # Pad to the nearest 4 byte boundary.
    printf -v "${pad}" "$(( (4 - ("${!len}" % 4)) % 4 ))"
    # Generate a format string for the variable.
    str="%%s\\\x00"
    for p in $(seq 1 "${!pad}") ; do
      str+="\\\x00"
    done
    printf -v "${fmt}" "${str}"
  done

  # Clean up the output file.
  rm -f "${output}"

  # Write the sizes for the ELF note name and description.
  for n in "${name_len}" "${input_len}" 0 ; do
    printf -v sz '\\x%02x\\x%02x\\x%02x\\x%02x' \
      $((n & 255)) $((n >> 8 & 255)) $((n >> 16 & 255)) $((n >> 24 & 255))
    printf "${sz}" >> "${output}"
  done

  # Write the ELF note name and description.
  printf "${name_fmt}" "${name}" >> "${output}"
  printf "${input_fmt}" "${input}" >> "${output}"
}
```

Reply:
good catch! just published a new version that uses this script with one modification: after we calculate the padding size I added that to the overall size of the section. without this, the parser can't properly find the start of the `desc` section

Reply:
> good catch! just published a new version that uses this script with one modification: after we calculate the padding size I added that to the overall size of the section. without this, the parser can't properly find the start of the desc section

The linked reference says:
> The "Name Size" and "Desc Size" fields are integers (in the byte order specified by the binary's ELF header) which specify the size of the "Name" and "Desc" fields (excluding padding).

So I don't think it's correct to add the padding to these size fields.

Reply:
I expect this is the right fix on the parsing side:
```
let mut shdr: SectionHeader = ... ;
shdr.sh_addralign = 4;
```

It looks like the address alignment defaults to one byte, which doesn't seem correct for note sections ([source](https://docs.oracle.com/cd/E23824_01/html/819-0690/chapter6-18048.html)):
> For 64–bit objects and 32–bit objects, each entry is an array of 4-byte words in the format of the target processor.
> ...
> Padding is present, if necessary, to ensure 4-byte alignment for the descriptor. Such padding is not included in namesz.

Reply:
ah ok, noted. removed my change and will fix the parsing logic